drivers/crypto/crypto_rpi_pico_sha256.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2025 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_sha256
+
+#include <zephyr/crypto/crypto.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <pico/bootrom/lock.h>
+#include <pico/sha256.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sha_rpi_pico_sha256, CONFIG_CRYPTO_LOG_LEVEL);
+
+struct rpi_pico_sha256_data {
+	pico_sha256_state_t state;
+};
+
+static int rpi_pico_sha256_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	struct rpi_pico_sha256_data *data = ctx->device->data;
+
+	assert(data->state.locked);
+
+	data->state.cache_used = 0;
+	data->state.cache.word = 0;
+	data->state.total_data_size = 0;
+
+	sha256_err_not_ready_clear();
+	sha256_set_bswap(true);
+	sha256_start();
+
+	pico_sha256_update(&data->state, pkt->in_buf, pkt->in_len);
+
+	if (!finish) {
+		return 0;
+	}
+
+	pico_sha256_write_padding(&data->state);
+	sha256_wait_valid_blocking();
+
+	for (uint i = 0; i < 8; i++) {
+		((uint32_t *)pkt->out_buf)[i] = BSWAP_32((uint32_t)sha256_hw->sum[i]);
+	}
+
+	return 0;
+}
+
+static int rpi_pico_sha256_query_hw_caps(const struct device *dev)
+{
+	return (CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
+}
+
+static int rpi_pico_sha256_hash_begin_session(const struct device *dev, struct hash_ctx *ctx,
+					      enum hash_algo algo)
+{
+	struct rpi_pico_sha256_data *data = dev->data;
+
+	assert(!data->state.locked);
```

Comment:
Isn't `assert()` a bit too strong here? I suggest adding a mutex in `struct rpi_pico_sha256_data` and use it here:
- lock the mutex
- check `data->state.locked` and return if already set (releasing the mutex, of course before returning)
- set `data->state.locked` (not on line 79 as it is done now)
- unlock the mutex
- remember to unset `data->state.locked` in `rpi_pico_sha256_hash_session_free()`

In this way you can have 2 threads willing to use the hash driver at the same time and then the one that comes later will just wait for the resource to be free.